src/Proto3/Suite/DotProto/Generate.hs

```diff
@@ -868,17 +868,18 @@ messageInstD ctxt parentIdent msgIdent messageParts = do
                       --    Constructor y -> encodeMessageField num (Nested (Just y)) -- for embedded messages
                       --    Constructor y -> encodeMessageField num (ForceEmit y)     -- for everything else
                       let mkAlt (OneofSubfield fieldNum conName _ dpType options) = do
-                            let wrapMaybe
+                            let isMaybe
                                    | Prim (Named tyName) <- dpType
-                                   , Just DotProtoKindMessage <- dotProtoTypeInfoKind <$> M.lookup tyName ctxt
-                                   = HsParen . HsApp (HsVar (haskellName "Just"))
+                                   = Just DotProtoKindMessage == fmap dotProtoTypeInfoKind (M.lookup tyName ctxt)
                                    | otherwise
-                                   = forceEmitE
+                                   = False
 
-                            xE <- wrapE ctxt options dpType
-                                   . wrapMaybe
-                                   $ HsVar (unqual_ "y")
+                            let wrapJust = HsParen . HsApp (HsVar (haskellName "Just"))
 
+                            xE <- (if isMaybe then id else fmap forceEmitE)
+                                   . wrapE ctxt options dpType
```

Comment:
Using `dimap` would be nice here.

Reply:
Sorry, I don't see how to use dimap at all. Any hints?

Reply:
Oops, you're right -- I thought both the pre- and post- operations were enabled by `isMaybe`.

This might read better though:

```
let adapt | isMaybe = lmap wrapJust
                | otherwise = rmap (fmap forceEmitE)

xE <- adapt (wrapE ctxt options dpType) (HsVar (unqual_ "y"))
```

Reply:
I find that harder to read. I guess I'd usually `(. wrapJust)` and `(fmap forceEmitE .)` rather than lmap/rmap - and in this case I'm not sure it's clearer at all. But happy to go that way if you want.

Reply:
I was thinking that it would be better to consolidate the adjustments predicated on `isMaybe`, but on second thought the extra indirection obfuscates matters.